src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs

```diff
@@ -0,0 +1,313 @@
+    {
+        var length = value.Length;
+
+        byte[] buffer = new byte[length + 2];
+
+        Encoding.ASCII.GetBytes(value, 0, length, buffer, 2);
+
+        MessagePackSerializer.WriteStr8Header(buffer, 0, length);
+
+        return buffer;
+    }
+
+    // This method would map the logger category to a table name which only contains alphanumeric values with the following additions:
+    // Any character that is not allowed will be removed.
+    // If the resulting string is longer than 50 characters, only the first 50 characters will be taken.
+    // If the first character in the resulting string is a lower-case alphabet, it will be converted to the corresponding upper-case.
+    // If the resulting string still does not comply with Rule, the category name will not be serialized.
+    private static int SanitizedCategoryName(Span<byte> buffer, string categoryName)
+    {
+        // Reserve 2 bytes for storing LIMIT_MAX_STR8_LENGTH_IN_BYTES and (byte)validNameLength -
+        // these 2 bytes will be back filled after iterating through categoryName.
+        int cursor = 2;
+        int validNameLength = 0;
+
+        // Special treatment for the first character.
+        var firstChar = categoryName[0];
+        if (firstChar >= 'A' && firstChar <= 'Z')
+        {
+            buffer[cursor++] = (byte)firstChar;
+            ++validNameLength;
+        }
+        else if (firstChar >= 'a' && firstChar <= 'z')
+        {
+            // If the first character in the resulting string is a lower-case alphabet,
+            // it will be converted to the corresponding upper-case.
+            buffer[cursor++] = (byte)(firstChar - 32);
+            ++validNameLength;
+        }
+        else
+        {
+            // Not a valid name.
+            return 0;
+        }
+
+        for (int i = 1; i < categoryName.Length; ++i)
+        {
+            if (validNameLength == MaxSanitizedEventNameLength)
+            {
+                break;
+            }
+
+            var cur = categoryName[i];
+            if ((cur >= 'a' && cur <= 'z') || (cur >= 'A' && cur <= 'Z') || (cur >= '0' && cur <= '9'))
+            {
+                buffer[cursor++] = (byte)cur;
+                ++validNameLength;
+            }
+        }
+
+        // Backfilling MessagePack serialization protocol and valid category length to the startIdx of the categoryName byte array.
+        MessagePackSerializer.WriteStr8Header(buffer, 0, validNameLength);
+
+        return cursor;
+    }
+
+    private byte[] ResolveTableMappingForCategoryName(string categoryName)
+    {
+        var tableNameCache = this.m_tableNameCache;
+
+        if (tableNameCache.TryGetValue(categoryName, out byte[] tableName))
+        {
+            return tableName;
+        }
+
+        return this.ResolveTableMappingForCategoryNameRare(categoryName);
+    }
+
+    private byte[] ResolveTableMappingForCategoryNameRare(string categoryName)
+    {
+        lock (this.m_lockObject)
```

Comment:
Would a read-write-lock fit better here?

Reply:
Could we use read-write-lock or concurrent dictionary (as @utpilla suggested below) here? Yes. BUT... the performance is not identical. My goal here is to not force a perf hit on users happy without the prefix feature, if possible. My thinking with this design was based on the idea that if we took a limit to infinity, there are no writes to the dictionary at all 🤣 The writes only exist as a blip during warmup. So my goal was to keep the reads as fast as possible once everything is warm. With concurrent dictionary and `ReaderWriterLockSlim` there seems to be overhead. `Hashtable` is really close and avoids the need for copy-on-write. I'm considering making a change to use `Hashtable`.

|               Method |      Mean |     Error |    StdDev |
|--------------------- |----------:|----------:|----------:|
|           Dictionary |  9.505 ns | 0.2339 ns | 0.2188 ns |
| ConcurrentDictionary | 15.491 ns | 0.3507 ns | 0.3281 ns |
|           ReaderLock | 26.550 ns | 0.5548 ns | 0.7595 ns |
|            Hashtable | 10.890 ns | 0.2053 ns | 0.1920 ns |

<details>
<summary>Benchmark code</summary>

```csharp
using System;
using System.Collections;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Threading;
using BenchmarkDotNet.Attributes;

namespace OpenTelemetry.Exporter.Geneva.Benchmark;

public class DictionaryBenchmarks
{
    private readonly ReaderWriterLockSlim readerWriterLock = new();

    private readonly Dictionary<string, byte[]> values = new(StringComparer.Ordinal)
    {
        ["Category1"] = new byte[1],
        ["Category2"] = new byte[1],
        ["Category3"] = new byte[1],
    };

    private readonly ConcurrentDictionary<string, byte[]> cvalues = new(StringComparer.Ordinal)
    {
        ["Category1"] = new byte[1],
        ["Category2"] = new byte[1],
        ["Category3"] = new byte[1],
    };

    private readonly Hashtable hvalues = new(StringComparer.Ordinal)
    {
        ["Category1"] = new byte[1],
        ["Category2"] = new byte[1],
        ["Category3"] = new byte[1],
    };

    [Benchmark]
    public byte[] Dictionary()
    {
        if (this.values.TryGetValue("Category1", out var value))
        {
            return value;
        }

        lock (this.values)
        {
            return LookupCategoryNameRare("Category1");
        }
    }

    [Benchmark]
    public byte[] ConcurrentDictionary()
    {
        return this.cvalues.GetOrAdd("Category1", LookupCategoryNameRare);
    }

    [Benchmark]
    public byte[] ReaderLock()
    {
        this.readerWriterLock.EnterReadLock();
        try
        {
            if (this.values.TryGetValue("Category1", out var value))
            {
                return value;
            }
        }
        finally
        {
            this.readerWriterLock.ExitReadLock();
        }

        this.readerWriterLock.EnterWriteLock();
        try
        {
            return LookupCategoryNameRare("Category1");
        }
        finally
        {
            this.readerWriterLock.ExitWriteLock();
        }
    }

    [Benchmark]
    public byte[] Hashtable()
    {
        byte[] mapping = (byte[])this.hvalues["Category1"];
        if (mapping != null)
        {
            return mapping;
        }

        lock (this.hvalues)
        {
            return LookupCategoryNameRare("Category1");
        }
    }

    private static byte[] LookupCategoryNameRare(string categoryName)
    {
        throw new NotSupportedException("All the benchmarks should hit the cache.");
    }
}
```
</details>

Reply:
> My goal here is to not force a perf hit on users happy without the prefix feature, if possible.

That's a good point.